oio/account/backend.py

```diff
@@ -913,14 +980,36 @@ def refresh_bucket(self, bucket_name, **kwargs):
         Refresh the counters of a bucket. Recompute them from the counters
         of all shards (containers).
         """
+        lkey = self.blockkey(bucket_name)
+        try:
+            ctime = int(time())
+            self.script_get_lock_bucket(keys=[lkey], args=[ctime])
+        except redis.exceptions.ResponseError as exc:
+            if six.text_type(exc).endswith("bucket_lock"):
+                raise Conflict("Refresh on bucket already in progress")
+            raise
+
         account_id = '__account__'
+        marker = ""
         keys = [AccountBackend.ckey(account_id, ''),
                 self.clistkey(account_id), self.bkey(bucket_name),
-                self.blistkey(account_id)]
+                marker]
         args = [bucket_name]
+
         try:
-            self.script_refresh_bucket(
-                keys=keys, args=args)
+            while True:
+                keys = [AccountBackend.ckey(account_id, ''),
+                        self.clistkey(account_id), self.bkey(bucket_name),
+                        marker, lkey]
+                res = self.script_refresh_bucket(keys=keys, args=args)
+                if not res[0]:
+                    break
+                marker = res[0]
+                # could be done in LUA refresh
+                self.conn.hset(lkey,
+                               mapping={"marker": marker,
+                                        "mtime": int(time())})
+            self.conn.delete(lkey)
```

Comment:
The lock should be released directly in the `refresh_bucket` script so as not to forget to increment new containers that are greater than the last marker.
Therefore, we should also take the lock directly from the `refresh_bucket` script.

Reply:
I don't get your point but I'll move `self.conn.delete(lkey)` in a `finally:` block to be sure that lock is properly removed.

Reply:
With an example, it will be clearer:
- the `refresh_bucket` script ended with the last `toto` marker
- the transaction is complete, so new requests can be executed on the master
- a new `tyty` container is created
- `"tyty" > "toto"`, so the bucket stats are not updated
- the lock key is deleted
- the bucket stats are not correct

Another solution can be to set a fake marker (`"\xef\xbf\xbd"`) at the end of the script when the refresh is finished.

Reply:
Gotcha !